File: app/src/main/java/me/manabu/activities/LoginActivity.java
package me.manabu.activities;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.Task;

import me.manabu.R;
import me.manabu.utils.AuthUtils;

public class LoginActivity extends AppCompatActivity implements View.OnClickListener {

    public static final int RC_ACTIVITY_LOGIN = 100;
    private static final int RC_LOGIN_GOOGLE = 101;

    private boolean doubleBackToExitPressedOnce = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        Button signInButton = (Button) findViewById(R.id.login_button_google);
        signInButton.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.login_button_google:
                Intent signInIntent = AuthUtils.getClientForActivity(this).getSignInIntent();
                startActivityForResult(signInIntent, RC_LOGIN_GOOGLE);
                break;
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        switch (resultCode){
            case RESULT_OK:
                break;
            case RESULT_CANCELED:
                return;

            default:
                Log.d("LoginActivity", "onActivityResult: result code check failed.");
                return;
        }

        switch (requestCode) {
            case RC_LOGIN_GOOGLE:
                Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
                handleSignInResult(task);
                break;
        }
    }

    private void handleSignInResult(Task<GoogleSignInAccount> completedTask) {
        try {
            GoogleSignInAccount account = completedTask.getResult(ApiException.class);
            AuthUtils.signIn(account);
            Log.d("LoginActivity", "handleSignInResult done.");
            setResult(RESULT_OK);
            finish();
        } catch (ApiException e) {
            Toast.makeText(this,
                    getString(R.string.login_error_google_sign_in) + ", #" + e.getStatusCode(),
                    Toast.LENGTH_LONG).show();
            Log.d("LoginActivity", "handleSignInResult failed: " + e.getStatusCode());
        }
    }

    @Override
    public void onBackPressed() {
        if (doubleBackToExitPressedOnce) {
            super.onBackPressed();
            return;
        }

        this.doubleBackToExitPressedOnce = true;
        Toast.makeText(this, getResources().getString(R.string.main_double_exit), Toast.LENGTH_SHORT).show();

        new Handler().postDelayed(() -> doubleBackToExitPressedOnce = false, 3000);
    }
}

File: app/src/main/java/me/manabu/activities/LoginActivity.kt
package me.manabu.activities

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.Toast
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.tasks.Task
import kotlinx.android.synthetic.main.activity_login.*
import me.manabu.R
import me.manabu.modules.CurrentUser
import me.manabu.modules.GoogleAuth
import org.jetbrains.anko.act
import org.jetbrains.anko.sdk25.coroutines.onClick
import org.jetbrains.anko.toast

class LoginActivity : AppCompatActivity() {

    companion object {
        const val RC_ACTIVITY_LOGIN = 100
        private const val RC_LOGIN_GOOGLE = 101
    }

    private var doubleBackToExitPressedOnce = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        loginButtonGoogle.onClick {
            val signInIntent = GoogleAuth.getClientForActivity(act).signInIntent
            startActivityForResult(signInIntent, RC_LOGIN_GOOGLE)
         }
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)

        if(resultCode != RESULT_OK){
            Log.d("LoginActivity", "onActivityResult: result code check failed.")
            return
        }

        when (requestCode) {
            RC_LOGIN_GOOGLE -> {
                val task = GoogleSignIn.getSignedInAccountFromIntent(data)
                handleSignInResult(task)
            }
        }
    }

    private fun handleSignInResult(completedTask: Task<GoogleSignInAccount>) {
        try {
            val account = completedTask.getResult(ApiException::class.java)
            CurrentUser.set(account)
            setResult(RESULT_OK, intent)
            finish()
        } catch (e: ApiException) {
            toast("${getString(R.string.login_error_google_sign_in)}, #${e.statusCode}")
        }

    }

    override fun onBackPressed() {
        if (doubleBackToExitPressedOnce) {
            super.onBackPressed()
            finish()
            return
        }

        this.doubleBackToExitPressedOnce = true
        toast(resources.getString(R.string.main_double_exit))

        Handler().postDelayed({ doubleBackToExitPressedOnce = false }, 3000)
    }
}